memory_debugging_tensor.py

```diff
@@ -0,0 +1,82 @@
+import torch
+from torch.utils._pytree import tree_map
+from torch.testing._internal.common_utils import TestCase, run_tests
+
+import weakref
+alive_tensors = weakref.WeakValueDictionary()
+
+# The main idea behind this tensor is to keep track of what tensors have been
+# allocated and track memory allocation through the course of a function.
+# Initially inspired by https://github.com/pytorch/pytorch/issues/72450, where I
+# wanted to understand *why* PyTorch had a peak memory usage of ~5x the original
+# tensor.
+
+class MemoryDebugTensor(torch.Tensor):
+    elem: torch.Tensor
+
+    __slots__ = ['elem']
+
+    @staticmethod
+    def __new__(cls, elem, func=None):
+        r = torch.Tensor._make_wrapper_subclass(
+            cls, elem.size(),
+            strides=elem.stride(), storage_offset=elem.storage_offset(),
+            # TODO: clone storage aliasing
+            dtype=elem.dtype, layout=elem.layout,
+            device=elem.device, requires_grad=elem.requires_grad
+        )
+        r.elem = elem
```

Comment:
This one shouldn't need the wrapper treatment, amirite?

Reply:
Uh... maybe, but that's how I'm used to doing it :^)